common/src/main/java/org/apache/cassandra/sidecar/common/JmxClient.java

```diff
@@ -0,0 +1,237 @@
+     *
+     * @param host the host of the JMX service
+     * @param port the port of the JMX service
+     */
+    public JmxClient(String host, int port)
+    {
+        this(host, port, null, null, false);
+    }
+
+    /**
+     * Creates a new client with the provided parameters
+     *
+     * @param host      the host of the JMX service
+     * @param port      the port of the JMX service
+     * @param role      the JMX role used for authentication
+     * @param password  the JMX role password used for authentication
+     * @param enableSSl true if SSL is enabled for JMX, false otherwise
+     */
+    public JmxClient(String host, int port, String role, String password, boolean enableSSl)
+    {
+        this(buildJmxServiceURL(host, port), role, password, enableSSl);
+    }
+
+    @VisibleForTesting
+    JmxClient(JMXServiceURL jmxServiceURL)
+    {
+        this(jmxServiceURL, null, null, false);
+    }
+
+    @VisibleForTesting
+    JmxClient(JMXServiceURL jmxServiceURL, String role, String password)
+    {
+        this(jmxServiceURL, role, password, false);
+    }
+
+    private JmxClient(JMXServiceURL jmxServiceURL, String role, String password, boolean enableSsl)
+    {
+        this.jmxServiceURL = jmxServiceURL;
+
+        jmxEnv = new HashMap<>();
+        if (role != null && password != null)
+        {
+            String[] credentials = new String[]{ role, password };
+            jmxEnv.put(JMXConnector.CREDENTIALS, credentials);
+        }
+        jmxEnv.put(REGISTRY_CONTEXT_SOCKET_FACTORY, getRMIClientSocketFactory(enableSsl));
+    }
+
+    /**
+     * Applies a function through the JMX bean proxy
+     *
+     * @param clientClass the management interface that the MBean exports, which will
+     *                    also be implemented by the returned proxy
+     * @param remoteName  the name of the MBean within {@code connection} to forward to
+     * @param func        the remote function that will be executed through the proxy
+     * @param <C>         the type of the proxy client
+     * @param <R>         the type of the result returned by the proxy
+     * @return the result of the remote function after being executed via the proxy
+     */
+    public <C, R> R apply(Class<C> clientClass, String remoteName, Function<C, R> func)
+    {
+        checkConnection();
+        C client = getProxy(clientClass, remoteName);
+        return func.apply(client);
+    }
+
+    /**
+     * Calls a remote function through the JMX bean proxy
+     *
+     * @param clientClass the management interface that the MBean exports, which will
+     *                    also be implemented by the returned proxy
+     * @param remoteName  the name of the MBean within {@code connection} to forward to
+     * @param caller      the consumer to be called through the proxy
+     * @param <C>         the type of the proxy client
+     */
+    public <C> void call(Class<C> clientClass, String remoteName, Consumer<C> caller)
```

Comment:
It is a bit counter-intuitive that `call` has no return value. I am thinking of `Callable`.

Similarly, `client.apply` reads not well. Semantically, `apply` reminds me of function evaluation. But `JmxClient` is not a function, instead it is a client to invoke/call a remote procedure.

How about remove both method and make `getProxy` public? This one the chained method call also looks fluent, and it no longer take a lambda, which can grow complex and look messy. For example,
```java
        jmxClient.getProxy(StorageJmxOperations.class, STORAGE_SERVICE_OBJ_NAME)
                 .takeSnapshot(tag, options, keyspace + "." + table);
```
Remember to call `checkConnection()` in `getProxy`, so it is identical to `apply` and `call`.

Reply:
trying out how this would look like, and I agree that the call looks fluent. Now the question is whether exposing the proxy is okay or not. I think since we expose it to internal classes, it should be fine. The proxy will be consumed directly by the interfacing class between the endpoint and Cassandra, so this would limit the exposure of the proxy.